File: iec61131lang/src/main/java/edu/kit/iti/formal/automation/datatypes/AnySignedInt.java
package edu.kit.iti.formal.automation.datatypes;

/*-
 * #%L
 * iec61131lang
 * %%
 * Copyright (C) 2016 Alexander Weigl
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

import java.math.BigInteger;

/**
 * <p>AnySignedInt class.</p>
 *
 * @author weigl
 * @version $Id: $Id
 */
public class AnySignedInt extends AnyInt {
    /**
     * <p>Constructor for AnySignedInt.</p>
     *
     * @param bits a int.
     */
    public AnySignedInt(int bits) {
        super(bits, true);
    }


    /**
     * {@inheritDoc}
     */
    @Override
    public AnyInt asSigned() {
        return this;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public AnyUnsignedInt asUnsgined() {
        return new AnyUnsignedInt(bitLength);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public AnyInt next() {
        return null;//TODO
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public boolean isValid(long value) {
        long max = (2 << bitLength) - 1;
        long min = -(2 << bitLength);
        return value <= max && value >= min;
    }

    @Override
    public BigInteger getUpperBound() {
        return BigInteger.ONE.shiftLeft(bitLength - 1).subtract(BigInteger.ONE);
    }

    @Override
    public BigInteger getLowerBound() {
        return BigInteger.ONE.shiftLeft(bitLength - 1).negate();
    }
}

File: iec61131lang/src/main/kotlin/edu/kit/iti/formal/automation/datatypes/AnySignedInt.kt
package edu.kit.iti.formal.automation.datatypes

/*-
 * #%L
 * iec61131lang
 * %%
 * Copyright (C) 2016 Alexander Weigl
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

import java.math.BigInteger
import java.util.Optional

/**
 *
 * AnySignedInt class.
 *
 * @author weigl
 * @version $Id: $Id
 */
abstract class AnySignedInt
/**
 *
 * Constructor for AnySignedInt.
 *
 * @param bits a int.
 */
(bits: Int) : AnyInt(bits, true) {

    override val upperBound: BigInteger
        get() = BigInteger.ONE.shiftLeft(bitLength - 1).subtract(BigInteger.ONE)

    override val lowerBound: BigInteger
        get() = BigInteger.ONE.shiftLeft(bitLength - 1).negate()


    /**
     * {@inheritDoc}
     */
    override fun asSigned(): AnyInt {
        return this
    }

    /**
     * {@inheritDoc}
     */
    override fun isValid(value: Long): Boolean {
        val max = ((2 shl bitLength) - 1).toLong()
        val min = (-(2 shl bitLength)).toLong()
        return value <= max && value >= min
    }

    class Arbitrary(bitLength: Int) : AnySignedInt(bitLength) {

        override fun next(): Optional<AnyInt> {
            return Optional.empty()
        }

        override fun asUnsgined(): AnyUnsignedInt {
            return AnyUnsignedInt.Arbitrary(bitLength)
        }
    }
}
